incubator/sparkoperator/README.md

```diff
@@ -21,14 +21,23 @@ Note that you need to use the `--namespace` flag during `helm install` to specif
-| `operatorImageName`       | The name of the operator image                        | `gcr.io/spark-operator/spark-operator` |
-| `operatorVersion`         | The version of the operator to install                | `v2.3.1-v1alpha1-latest`               |
-| `sparkJobNamespace`       | K8s namespace where Spark jobs are to be deployed.    | `default`                              |
-| `createSparkJobNamespace` | Whether to create the Spark job namespace             | false                                  |
-| `enableWebhook`           | Whether to enable mutating admission webhook          | true                                   |
-| `enableMetrics`           | Whether to expose metrics to be scraped by Premetheus | true                                   |
+| Parameter                 | Description                                                  | Default                                |
+| ------------------------- | ------------------------------------------------------------ | -------------------------------------- |
+| `operatorImageName`       | The name of the operator image                               | `gcr.io/spark-operator/spark-operator` |
+| `operatorVersion`         | The version of the operator to install                       | `v2.4.0-v1alpha1-latest`               |
+| `imagePullPolicy`         | Docker image pull policy                                     | `IfNotPresent`                         |
+| `sparkJobNamespace`       | K8s namespace where Spark jobs are to be deployed            | `default`                              |
+| `createSparkJobNamespace` | Whether to create the Spark job namespace                    | false                                  |
+| `enableWebhook`           | Whether to enable mutating admission webhook                 | true                                   |
+| `enableMetrics`           | Whether to expose metrics to be scraped by Premetheus        | true                                   |
+| `controllerThreads`       | Number of worker threads used by the SparkApplication controller | 10                                     |
+| `installCrds`             | Whether to install CRDs                                      | true                                   |
+| `metricsPort`             | Port for the metrics endpoint                                | 10254                                  |
+| `metricsEndpoint`         | Metrics endpoint                                             | "/metrics"                             |
+| `metricsPrefix`           | Prefix for the metrics                                       | ""                                     |
+| `namespace`               | The Kubernetes namespace to manage. Will manage custom resource objects of the managed CRD types for the whole cluster if set to empty string. | ""                                     |
```

Comment:
This is somehow conflicting with `sparkJobNamespace`. Applications will be submitted to run in the namespace of their corresponding `SparkApplication`s. If this is set, then setting `sparkJobNamespace` separately (to a different value) makes no sense. If this is not set, setting `sparkJobNamespace` makes no sense either as users may create `SparkApplication`s in any namespace. It looks to me that we should get rid of `sparkJobNamespace` as the whole point of having it seems to setup the RBAC setting for the driver pods in that namespace.

Reply:
My understanding is that the point of `sparkJobNamespace` is to give the user a namespace that has RBAC pre-configured so that he can deploy jobs there out of the box. He can obviously choose to deploy jobs in other namespaces and have them managed by the operator (i.e. the `namespace` setting), knowing that the he needs to manually configure RBAC in those places. So I don't think these two settings are necessarily in conflict. But they are more like independent to me.

Reply:
They are not independent. If `namespace` is set, then setting `sparkJobNamespace` to anything other than `namespace` renders the namespace referenced by `sparkJobNamespace` useless as `SparkApplications` in that namespace won't be handled and run. I suggest we combine the two together into a single `namespace` parameter.

Reply:
Oh right I see what you meant. But if we have a single combined parameter, there wouldn't be any namespace that has RBAC pre-configured when the user wants the operator to monitor all namespaces. In other words, having two parameters allows the operator to monitor all namespaces, while at the same time set up RBAC in a namespace that the user can use out of the box without extra work. What do you think?

Reply:
When you have both, and set them to different values, the pre-configured RBAC for `sparkJobNamespace` is effectively useless because the operator won't watch and process applications from that namespace.

Reply:
Ok makes sense. I'll combine them then.

Reply:
@yuchaoran2011 have you got a chance to combine them?

Reply:
Sorry about the delay. I was tied up in other things. I'm planning to get the webhook fix in first. Then I'll close this PR, re-work it and make another one. Because this one seems to be stuck at the CI failure. I'm not sure adding new changes here would help.